### DISCUSSION TITLE
avoid one-off prepared statements

### DISCUSSION THREAD

**Discussion by gdarbaidze** [sequelize/sequelize#16365] [Help & Questions]
I have a problem with Sequelize(mysql dialect) and AWS RDS proxy.
The proxy gives logs like this:
`The client session was pinned to the database connection [dbConnection=3372749374] for the remainder of the session. The proxy can't reuse this connection until the session ends. Reason: A protocol-level prepared statement was detected.`
Meaning, that the proxy can't share connections, that execute prepared statements with other clients.

For some reason, Create method of a model generates prepared statements instead of injecting values into SQL query, which causes inefficient utilization of proxy's resources.
I discovered, that Sequelize have undocumented "bindParam" option, if you set bindParam: false in configuration, it starts to use value replacements instead of prepared statements.

But, it is safe to use "bindParam" for that?  
Is there any safe way to avoid prepared statements?

Related issue:
[https://github.com/sequelize/sequelize/issues/13972](https://github.com/sequelize/sequelize/issues/13972)

**Comment by ephys** [comment]
It is fine to use bindParam for now. We plan on providing a way to specify which queries use bind parameters and which use replacements, but that is not currently implemented

**Reply by KristoferBorgware** [reply]
Is there any update for this, or is it still considered a longer-term idea?

This would be particularly helpful for setups running behind **AWS RDS Proxy**, where controlling prepared-statement behavior can have a noticeable impact on connection pinning.

Thanks — and appreciate all the work on Sequelize.

**Comment by adixists** [comment]
Hey,
I ran into this exact same issue with AWS RDS Proxy + Sequelize (MySQL) a few months ago. The "protocol-level prepared statement" pinning is super annoying and kills connection reuse.
What I Learned
You're right — Sequelize defaults to using prepared statements (? placeholders + bindParam) with the MySQL driver for safety (prevents SQL injection). But RDS Proxy sees that and pins the connection for the rest of the session, which defeats the whole purpose of the proxy.
The bindParam: false trick you found does work to force value interpolation instead of prepared statements. However, you're right to be cautious.
My Recommendation (What Worked for Me)
"
const sequelize = new Sequelize({
  dialect: 'mysql',
  ...,
  dialectOptions: {
    // This is the key
    bindParam: false,
  }
});
"
Is it safe?

Yes, as long as you're using Sequelize's model methods properly (Model.create(), Model.update(), etc.) and not manually building raw queries with user input.
Sequelize still escapes values on its side before injecting them, so SQL injection risk is very low in normal usage.
I’ve been running this in production for ~4 months now without issues.

Alternative Options

Best long-term fix → Use RDS Proxy with connection pooling disabled for prepared statements (not ideal) or switch to Aurora + native driver pooling if possible.
Use sequelize.literal() carefully only where needed.
Some people disable prepared statements only for specific heavy-read models.

Would love to hear if setting bindParam: false globally solved your pinning problem completely. Also, if you can share a bit more about your workload (read-heavy? lots of creates?), maybe we can find a more targeted solution.
